### PR TITLE
Allow users to just assign mount points to block devices

### PR DESCRIPTION
**Pull request by vpodzime**
This PR adds a new way of storage configuration to the Anaconda installer. It is the most trivial way of storage configuration possible -- just assigning mount points to block devices and optionally reformatting them. Users are allowed and encouraged to do whatever they want with their storage using whichever tools they want eventually just telling the installer which block device to mount where.

There are still certain limitations to keep things consistent, though. In particular, the file system mounted as ``/`` has to be reformatted as part of the installation process and all ``swap`` formats need to be reformatted in order to be used by the newly installed system.

The first three patches are quite well isolated to limit the potential regression surface. The fourth one is invasive, but on the other hand only making things safer and less vulnerable. Without it, the repeated visits of the storage configuration screens are doomed to fail (regardless of the new mount point assignment functionality). It's much safer to just start from scratch if users change their mind about how they want to configure storage.

I plan to implement a few kickstart tests for this new functionality if it gets merged. The goal here is to make this new feature available in Fedora 27 as a new (maybe experimental?) way of storage configuration in text mode.

Note: This requires changes from https://github.com/rhinstaller/pykickstart/pull/178 and a matching bump of the required version of *pykickstart*

### CODE REVIEW

File: pyanaconda/ui/tui/spokes/storage.py [hunk 1]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+        self._container = None
+        self._mds = None
+
+        self._gather_mount_data_info()
+
+    def _gather_mount_data_info(self):
+        self._mds = OrderedDict()
+
+        for device in (d for d in self.storage.devicetree.leaves if not d.protected and d.size != Size(0)):
+            if device.format.type != "none":
```
Comment (M4rtinK) [inline pyanaconda/ui/tui/spokes/storage.py]: Shouldn't there be a constant for this ?

File: pyanaconda/ui/tui/spokes/storage.py [hunk 2]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+                    devspec += "*"
+                if md.mount_point:
+                    devspec += ", %s" % md.mount_point
+            w = TextWidget(devspec)
+            self._container.add(w, self._configure_device, device)
+
+        self.window.add_with_separator(self._container)
+
+        message = _("Choose device from above to assign mount point and set format.\n" +
+                    "Formats marked with * are new formats meaning all DATA on the original format WILL BE LOST!")
```
Comment (M4rtinK) [inline pyanaconda/ui/tui/spokes/storage.py]: I would suggest making all also capital, eq.: **ALL** DATA

File: pyanaconda/ui/tui/spokes/storage.py [hunk 3]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+            elif key.lower() == C_('TUI|Spoke Navigation', 'c'):
+                self.apply()
+
+            return super(MountPointAssignSpoke, self).input(args, key)
+
+        return InputState.PROCESSED
+
+    def apply(self):
+        """ Apply our selections. """
+        for md in self._mds.values():
```
Comment (M4rtinK) [inline pyanaconda/ui/tui/spokes/storage.py]: What's actually `md` in this case ? I would suggest using a longer variable name for clarity.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 4]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+        NormalTUISpoke.refresh(self, args)
+
+        self._container = ListColumnContainer(1)
+
+        fmt = get_format(self._md.format)
+        if fmt and fmt.mountable:
+            self._container.add(TextWidget("Mount point: %s" % self._md.mount_point or "none"), self._assign_mount_point, self._md)
+        else:
+            self._container.add(TextWidget("Mount point: [%s]" % fmt.name), lambda x: self.redraw(), None)
+        self._container.add(TextWidget("Format: %s" % self._md.format or "none"), self._set_format, self._md)
```
Comment (M4rtinK) [inline pyanaconda/ui/tui/spokes/storage.py]: Same suggestion - I think `"none"` should be defined by a constant there as well.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 5]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+            return Prompt(_("New mount point for the device"))
+        else:
+            self.apply()
+            self.close()
+
+    def input(self, args, key):
+        if not key:
+            self.value = None
+            return InputState.DISCARDED
+        elif key.startswith("/"):
```
Comment (M4rtinK) [inline pyanaconda/ui/tui/spokes/storage.py]: I think a comment there might be helpful.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 6]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
     def apply(self):
         """ Apply our selections. """
         self.data.autopart.type = self._selected_scheme_value
+
+class MountDataRec(kickstart.MountData):
+    """ An artificial subclass also recording changes """
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Please write period on the end of sentences in docstrings.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 7]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+                self.data.mount.add_mount_data(md)
+
+class ConfigureDeviceSpoke(NormalTUISpoke):
+    """ Assign mount point to a block device and (optionally) reformat it """
+    category = SystemCategory
+
+    def __init__(self, data, storage, payload, instclass, mount_data):
+        NormalTUISpoke.__init__(self, data, storage, payload, instclass)
+        self._container = None
+        self._md = mount_data
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Can you please change the name from `_md` to `_mount_data`.

I don't like two latter variable for something used in the whole class. It could be anything like for example `MetaData`...

File: pyanaconda/ui/tui/spokes/storage.py [hunk 8]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+        if ((self._md.orig_format and self._md.orig_format != self._md.format) or self._md.mount_point == "/"):
+            # changing format implies reformat and so does "/" mount point
+            self._container.add(CheckboxWidget(title="Reformat", completed=self._md.reformat), lambda x: self.redraw(), None)
+        else:
+            self._container.add(CheckboxWidget(title="Reformat", completed=self._md.reformat), self._switch_reformat, self._md)
+
+        self.window.add_with_separator(self._container)
+
+        message = _("Choose from above to assign mount point and/or set format.")
+        self.window.add_with_separator(TextWidget(message))
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: I guess you can inline the `message` here. You have bigger lines above which are ok.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 9]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+        ScreenHandler.push_screen(dialog)
+
+class SetFormatDialog(EditTUIDialog):
+    """ Set format for a device """
+    category = SystemCategory
+
+    def __init__(self, data, storage, payload, instclass, mount_data):
+        EditTUIDialog.__init__(self, data, storage, payload, instclass)
+        self.title = N_("Configure device format")
+        self._md = mount_data
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: The same here.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 10]
```diff
@@ -576,3 +596,256 @@ def input(self, args, key):
...
+
+    @property
+    def indirect(self):
+        return True
+
+    def prompt(self, args=None):
+        if self.value is None:  # first run or nothing entered
+            return Prompt(_("New format for the device%s") % (" [%s]" % self._md.format if self._md.format else ""))
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: First of all it is not clear what you are doing when you have `%` within `%`.
Second I would move the space to the original string so that `...device%s` will be `...device %s`.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 11]
```diff
@@ -201,6 +201,13 @@ def _update_summary(self):
     def refresh(self, args=None):
         NormalTUISpoke.refresh(self, args)
 
+        if not any(d in self.storage.disks for d in self.disks):
+            # something happened to self.storage (probably reset), need to
+            # reinitialize the list of disks
+            threadMgr.add(AnacondaThread(name=THREAD_STORAGE_WATCHER,
+                                         target=self._initialize))
+
+
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Remove one empty line here.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 12]
```diff
@@ -367,7 +374,7 @@ def apply(self):
         if self.autopart:
             self.clearPartType = CLEARPART_TYPE_ALL
         else:
-            self.clearPartType = CLEARPART_TYPE_NONE
+            self.clearPartType = None
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Why this change? What the `None` here means?

File: pyanaconda/ui/tui/spokes/storage.py [hunk 13]
```diff
@@ -513,27 +526,51 @@ def apply(self):
...
+            self.data.clearpart.type = CLEARPART_TYPE_NONE
+            self.data.clearpart.initAll = False
+
+    def _ensure_init_storage(self):
+        """
+        If a different clearpart type was chosen or mount point assignment was
+        chosen instead, we need to reset/rescan storage to revert all changes
+        done by the previous run of doKickstartStorage() and get everything into
+        the initial state.
+
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Remove this empty line please.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 14]
```diff
@@ -616,10 +653,18 @@ def __init__(self, data, storage, payload, instclass):
 
         self._gather_mount_data_info()
 
+    def _is_dev_usable(self, dev):
+        maybe = not dev.protected and dev.size != Size(0)
+        if maybe and self.data.ignoredisk.onlyuse:
+            # all device's disks have to be in ignoredisk.onlyuse
+            maybe = set(self.data.ignoredisk.onlyuse).issuperset({d.name for d in dev.disks})
+
+        return maybe
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Isn't this code which should be in Blivet?

File: pyanaconda/ui/tui/spokes/storage.py [hunk 15]
```diff
@@ -513,27 +526,51 @@ def apply(self):
...
             self.data.autopart.autopart = True
             self.data.clearpart.type = self.clearPartType
             self.data.clearpart.initAll = True
             self.data.mount.clear_mount_data()
         else:
             self.data.autopart.autopart = False
+            self.data.clearpart.type = CLEARPART_TYPE_NONE
+            self.data.clearpart.initAll = False
+
+    def _ensure_init_storage(self):
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/storage.py]: Could this be done as common function somewhere else, maybe in Blivet directly.
This sounds to me like reset storage from any state to the original state, isn't this basically `Blivet.reset()`?

File: pyanaconda/kickstart.py [hunk 16]
```diff
@@ -1180,14 +1189,17 @@ def execute(self, storage, ksdata, instClass):
                                               msg=_("No format on device '%s'" % self.device)))
                 fmt = get_format(old_fmt.type)
             storage.format_device(dev, fmt)
+            # make sure swaps end up in /etc/fstab
+            if fmt.type == "swap":
+                storage.add_fstab_swap(dev)
 
         # only set mount points for mountable formats
-        if dev.format.mountable:
+        if dev.format.mountable and self.mount_point is not None and self.mount_point != "none":
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply pyanaconda/kickstart.py]: ``"none"``  can end up here from kickstart. See https://github.com/rhinstaller/pykickstart/pull/178/files#diff-626dd4fad4450cf40de2c467d3df3262R1163 for details.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 1 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: Yes, there should be a constant for this, but it should be defined in *Blivet*, not here. And I don't want to introduce a dependency on a new release of blivet just because of this.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 4 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: Why? It just says there is no format. Maybe it should be translated, though? Fixing.

**Comment by vpodzime** [comment]
@M4rtinK's comments addressed and the new feature is marked as EXPERIMENTAL. I'll take a look at the other concerns tomorrow.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 12 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: This change is a leftover from an older version. But you know what, the ``self.clearPartType`` attribute is only set in this class and never used. So I'm just removing it instead.

File: pyanaconda/ui/tui/spokes/storage.py [hunk 17]
```diff
@@ -450,48 +455,55 @@ def _initialize(self):
...
-       .. inheritance-diagram:: AutoPartSpoke
+       .. inheritance-diagram:: PartTypeSpoke
           :parts: 3
     """
     category = SystemCategory
 
     def __init__(self, data, storage, payload, instclass):
         NormalTUISpoke.__init__(self, data, storage, payload, instclass)
-        self.title = N_("Autopartitioning Options")
+        self.title = N_("Partitioning Options")
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: Yeah, i was thinking the same. Better than having a misleading title and it's text mode after all.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 15 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: ``storage_initialize()`` is a bit special beast which tries to make sure it behaves more friendly than ``reset()`` in the installation environment. And this method itself is very specific to text mode and users changing their mind. GUI uses a totally different approach.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 14 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: Not really, it says which devices to show on this screen.

File: pyanaconda/ui/tui/spokes/storage.py (hunk 10 above)
Comment (vpodzime) [reply pyanaconda/ui/tui/spokes/storage.py]: But that would mean there would always be a space after ``device`` which is exactly the result I want to prevent.

**Comment by vpodzime** [comment]
I believe all the comments should be addressed now.

**Comment by M4rtinK** [comment]
> I believe all the comments should be addressed now.
Yep, looks good to me! :) Once the Pykickstart PR is merged and a new build done I'll merge this and do a new Anaconda build as well + Bodhi.
